### DISCUSSION TITLE
Livewire requests will not be recognized as JSON requests

### DISCUSSION THREAD

**Discussion by ben182** [livewire/livewire#3531] [7. Bugs (archived)]
Hi,

I have this strange bug with Livewire, which I am observing for a while now. So, I have a PWA, which polls a couple of Livewire components on a regular basis. When I go from work to home or simply put my laptop into sleep and the PWA stays in background, I got this strange console error that basically crashes the whole Livewire script and disables any further network requests:

`
Uncaught (in promise) SyntaxError: Unexpected token < in JSON at position 0
    at JSON.parse (<anonymous>)
    at index.js:58
`

Sadly, I can not get any information on the network requests.
So, what this message obviously means is that Livewire (for whatever reason) receives some HTML from its request which is weird because of the Content-Type. Anyway, the line of code that triggers this error is this one: https://github.com/livewire/livewire/blob/4720f5abe8a9d1f89cec752497acbff09422a2d2/js/connection/index.js#L58

It tries to parse the received JSON from the request made by Livewire. This does not work because it is HTML. So, what I would expect from Livewire in this scenario is to throw an error or give me some option to react on this. It should check for a valid JSON response. I am pretty much not able to debug this any deeper. It only happens on production and only when the browser goes to sleep for a while. Session lifetime etc. checked (but should not be the problem because Livewire checks for 419). So it has to be a valid request - 200 response etc.

**Comment by joshhanley** [comment]
@ben182 what I suspect is happening is you're not running into the session lifetime (which should 419), but actually some form of redirect (maybe to the login form, to re-log the user back in).

You could try and manually trigger this, by logging in, get your component running, open another window and log the user out, then try interacting with the LW component in the original window. If you put a breakpoint on that line, you will probably see the response is a full HTML response, rather than JSON. 

If it is that, I've seen it happen before. Not sure how best to handle it so it doesn't happen but something that can be looked into.

Hope this helps!

**Reply by ben182** [reply]
@joshhanley thanks for your quick reply. I spent some hours now to look more into it and the trick that you provided worked indeed to reproduce the issue. I invalided the session by logging the user out, and all the next requests fail. But they don't throw errors! They respond with a 302 and redirect the Livewire requests to the login route. This route then obviously responds with HTML. That's the issue at its core. So I dived deeper and looked why it responds with a 302 and not with a normal JSON 401. The reason is that Laravel thinks that each Livewire request is **not** a JSON request. In fact, Livewire explicitly sets its Accept Header to HTML. That's a major problem in my understanding. Why is that the case? It always responds with JSON, doesn't it? So when we would set the Accept header here to JSON, it would fix everything.
https://github.com/livewire/livewire/blob/4720f5abe8a9d1f89cec752497acbff09422a2d2/js/connection/index.js#L41

**Comment by ben182** [comment]
I found other people with related problems / workarounds:
https://github.com/livewire/livewire/discussions/3428
https://github.com/livewire/livewire/discussions/3125
https://github.com/livewire/livewire/issues/243
https://www.matthewerwin.co.uk/blog/protecting-livewire-components-when-using-multiple-auth-drivers/
https://forum.laravel-livewire.com/t/how-to-make-a-component-accessible-for-authenticed-users-only/3996

**Comment by ben182** [comment]
@calebporzio Is there a specific reason why Livewire sets its Accept Header to HTML but deals with JSON? Because otherwise I would open a PR to fix this.

**Reply by joshhanley** [reply]
@ben182 I'm not sure why it's set that way. If you have a PR already prepared, maybe submit it. That way, if there are issues with it, it will then be discussed 🙂
